src/main/java/org/opentripplanner/standalone/server/RequestTraceFilter.java

```diff
@@ -0,0 +1,134 @@
+
+  /**
+   * The regular Random is good enough for generating new correlationIds, but we want
+   * to properly seed it. Many otp instances might get started at the same time so using
+   * the current time is not good enough. We do not use UUID because an 32 bit int is
+   * enough and easier to read.
+   */
+  private static final Random ID_GEN = new Random(new SecureRandom().nextLong());
+
+  /**
+   * This can not be final since it is injected at startup time.
+   */
+  private static List<RequestTraceParameter> traceParameters;
+
+  public static void init(List<RequestTraceParameter> parameters) {
+    var copy = new ArrayList<>(parameters);
+    // Reverse list to give precedence to elements in the beginning of the list. If the same
+    // logKey or responseHeader is used more than one time the first element will over-write
+    // the second when saved in the response and in the MDC map.
+    Collections.reverse(copy);
+    traceParameters = List.copyOf(copy);
+  }
+
+  /**
+   * The all trance-parameters and read in from http-request header; Generate value if
+   * not present in request or header key does not exist. At last insert key/value into
+   * log context.
+   */
+
+  @Override
+  public void filter(ContainerRequestContext requestContext) throws IOException {
+    for (var it : traceParameters) {
+      String value = null;
+      if (it.hasHttpRequestHeader()) {
+        value = requestContext.getHeaderString(it.httpRequestHeader());
+      }
+      if (value == null && it.generateIdIfMissing()) {
+        value = generateUniqueId();
+      }
+      if (it.hasLogKey()) {
+        // Put value, even if it is empty. This will clear any value set in the thread from
+        // before. This happens if the previous request did not clean up after it-self.
+        LogMDCSupport.putLocal(it.logKey(), value);
+      }
+    }
+  }
+
+  /**
+   * Insert all trace-parameters in http response headers. The value is taken from the request
+   * header, or log context, or generated.
+   */
+  @Override
+  public void filter(
+    ContainerRequestContext requestContext,
+    ContainerResponseContext responseContext
+  ) throws IOException {
+    for (var it : traceParameters) {
+      if (it.hasHttpResponseHeader()) {
+        var value = resolveValue(it, requestContext);
+        setHttpResponseHeaderValue(responseContext, it.httpResponseHeader(), value);
+      }
+      if (it.hasLogKey()) {
+        LogMDCSupport.removeLocal(it.logKey());
+      }
+    }
+  }
+
+  /**
+   * Find the value for the trace parameter:
+   * <ol>
+   *   <li>from the request http header, or</li>
+   *   <li>from the log context, or</li>
+   *   <li>generate a new unique value.</li>
+   * </ol>
+   */
+  private String resolveValue(RequestTraceParameter p, ContainerRequestContext requestContext) {
+    String value = null;
+
+    if (p.hasHttpRequestHeader()) {
+      value = requestContext.getHeaderString(p.httpRequestHeader());
```

Comment:
should we sanitize the header value?

Reply:
What would be a good pattern/method to apply ? [^\c\s]{1,1000} (1 to 1000 characters and NOT control or space)?

Reply:
Looked into it, and you are allowed to pass in a lot, so I think we should not validate this. If we should then we should add a regular expression to the config for the trace parameter.

Reply:
I would say yes. This is sufficient to prevent log forging. Preventing all forms of script injections is probably out of scope here.